### DISCUSSION TITLE
Error (mismatched structure) when deserializing Transformer NER pipeline from bytes

### DISCUSSION THREAD

**Discussion by salman1993** [explosion/spaCy#7485] [Help: Coding & Implementations] [answered]
I have a trained a Transformer NER pipeline on a GPU machine and then copied over the trained model to my local machine (no GPU). I am trying to serialize the pipeline and then deserialize it from bytes_data as shown here - https://spacy.io/usage/saving-loading

However, I am running into this error: `ValueError: Cannot deserialize model: mismatched structure` 

Versions:
spacy: 3.0.5
spacy-transformers: 1.0.1

The same versions were used during training.  I also tried the same code to serialize/deserialize on the GPU machine but it did not work there either. I think it's related to the `spacy_transformers.configure_strided_spans..get_strided_spans` but not sure.

Here is the screenshot:
![Screen Shot 2021-03-18 at 11 01 21 AM](https://user-images.githubusercontent.com/7154255/111648093-6e1d5080-87d9-11eb-8d27-8d6d3ef7a5d8.png)

Also tried `pickle.dump` but it fails:
![Screen Shot 2021-03-18 at 11 24 17 AM](https://user-images.githubusercontent.com/7154255/111651377-8b9fe980-87dc-11eb-9fe1-f712c4c8575e.png)

Ultimately, my goal is to write the model as pickle object and then load it. I was thinking:
```
spacy_model_dict = {
  "model": bytes_data, 
  "lang": lang, 
  "pipeline_names": pipeline
}

# write spacy_model_dict as pickle

# load pickle from saved file 
# initialize nlp.blank with lang, add pipes and then load from bytes data 
```

**Comment by salman1993** [comment]
This is the entire config - output from `nlp.config`: 
```
{'paths': {'train': 'corpus/train/',
  'dev': 'corpus/dev/',
  'vectors': None,
  'init_tok2vec': None},
 'system': {'gpu_allocator': 'pytorch', 'seed': 0},
 'nlp': {'lang': 'en',
  'pipeline': ['transformer', 'ner'],
  'batch_size': 512,
  'disabled': [],
  'before_creation': None,
  'after_creation': None,
  'after_pipeline_creation': None,
  'tokenizer': {'@tokenizers': 'spacy.Tokenizer.v1'}},
 'components': {'transformer': {'factory': 'transformer',
   'max_batch_items': 4096,
   'model': {'@architectures': 'spacy-transformers.TransformerModel.v1',
    'name': 'roberta-base',
    'get_spans': {'@span_getters': 'spacy-transformers.strided_spans.v1',
     'window': 128,
     'stride': 96},
    'tokenizer_config': {'use_fast': True}},
   'set_extra_annotations': {'@annotation_setters': 'spacy-transformers.null_annotation_setter.v1'}},
  'ner': {'factory': 'ner',
   'model': {'@architectures': 'spacy.TransitionBasedParser.v2',
    'state_type': 'ner',
    'extra_state_tokens': False,
    'hidden_width': 64,
    'maxout_pieces': 2,
    'use_upper': True,
    'nO': None,
    'tok2vec': {'@architectures': 'spacy-transformers.TransformerListener.v1',
     'grad_factor': 1.0,
     'pooling': {'@layers': 'reduce_mean.v1'},
     'upstream': '*'}},
   'moves': None,
   'update_with_oracle_cut_size': 100}},
 'corpora': {'dev': {'@readers': 'spacy.Corpus.v1',
   'path': '${paths.dev}',
   'max_length': 0,
   'gold_preproc': False,
   'limit': 0,
   'augmenter': None},
  'train': {'@readers': 'spacy.Corpus.v1',
   'path': '${paths.train}',
   'max_length': 512,
   'gold_preproc': False,
   'limit': 0,
   'augmenter': {'@augmenters': 'spacy.lower_case.v1', 'level': 0.2}}},
 'training': {'accumulate_gradient': 1,
  'dev_corpus': 'corpora.dev',
  'train_corpus': 'corpora.train',
  'seed': '${system.seed}',
  'gpu_allocator': '${system.gpu_allocator}',
  'dropout': 0.1,
  'patience': 1600,
  'max_epochs': 0,
  'max_steps': 6000,
  'eval_frequency': 200,
  'frozen_components': [],
  'before_to_disk': None,
  'batcher': {'@batchers': 'spacy.batch_by_padded.v1',
   'discard_oversize': True,
   'size': 2000,
   'buffer': 256,
   'get_length': None},
  'logger': {'@loggers': 'spacy.ConsoleLogger.v1', 'progress_bar': True},
  'optimizer': {'@optimizers': 'Adam.v1',
   'beta1': 0.9,
   'beta2': 0.999,
   'L2_is_weight_decay': True,
   'L2': 0.01,
   'grad_clip': 1.0,
   'use_averages': False,
   'eps': 1e-08,
   'learn_rate': {'@schedules': 'warmup_linear.v1',
    'warmup_steps': 250,
    'total_steps': 20000,
    'initial_rate': 5e-05}},
  'score_weights': {'ents_per_type': None,
   'ents_f': 1.0,
   'ents_p': 0.0,
   'ents_r': 0.0}},
 'pretraining': {},
 'initialize': {'vectors': None,
  'init_tok2vec': '${paths.init_tok2vec}',
  'vocab_data': None,
  'lookups': None,
  'before_init': None,
  'after_init': None,
  'components': {},
  'tokenizer': {}}}
```

**Comment by adrianeboyd** [comment, marked as answer]
Hi, I can see how that section of the docs is confusing. This was closer to how things worked in v2 and we should update it more for v3. Please see Matt's explanation to a similar question here: https://github.com/explosion/spaCy/issues/7079#issuecomment-779527242

I'm not entirely sure why you're not just using `nlp.to_disk`? The most common workflow is to use `nlp.to_disk("/path/to/dir")` to save and `spacy.load("/path/to/dir")` to load. You can also package a model directory with `spacy package` to a pip-installable package just like `en_core_web_sm` so you can load from a package name rather than a directory, including extra code for custom registered functions that's loaded automatically with the package.

We've implemented tons of helper functions for (de)serialization so you usually shouldn't need to write your own. But if you'd really like to customize this, you can see the basic steps in `util.load_model_from_path`, which is what `spacy.load` calls under the hood. You'd want to load the model from the config and then use `from_bytes` instead of `from_disk`.

https://github.com/explosion/spaCy/blob/61472e7cb385c6ca578dce2f4301fb27666e058b/spacy/util.py#L358-L390

We'd recommend avoiding pickle if you can. You should be able to save all the required data+config using msgpack, e.g. with `srsly.msgpack_dumps` / `srsly.msgpack_loads`.

The updated version of the docs would look like this:

```python
### Serialize
config = nlp.config
bytes_data = nlp.to_bytes()

### Deserialize
lang_cls = spacy.util.get_lang_class(config["nlp"]["lang"])
nlp = lang_cls.from_config(config)
nlp.from_bytes(bytes_data)
```
